### PR TITLE
Refactor Stateful operator and custom op

### PR DESCRIPTION
**Pull request by piiswrong**
@tqchen @eric-haibin-lin 
- Refactored CreateLayerOp to return any instead of Operator*.
- Use nnvm interface instead of Operator::Forward/Backward
- Added kLocal exec_type to run operators in main thread without pushing to engine.
- Refactored Custom op to use new interface

### CODE REVIEW

File: src/nnvm/legacy_op_util.cc [hunk 1]
```diff
@@ -328,10 +424,14 @@ void RegisterLegacyOpProp() {
     op.set_attr<nnvm::FMutateInputs>("FMutateInputs", OpPropMutateInputs);
     op.set_attr<nnvm::FInplaceOption>("FInplaceOption", OpPropInplaceOption);
     op.set_attr<FResourceRequest>("FResourceRequest", OpPropResourceRequest);
-    op.set_attr<FCreateLayerOp>("FCreateLayerOp", OpPropCreateLayerOp);
+    op.set_attr<FExecType>("FExecType", OpExecType);
+    op.set_attr<FCreateOpState>("FCreateOpState", OpPropCreateLayerOp);
+    op.set_attr<FStatefulCompute>("FStatefulCompute<cpu>", LegacyOpForward);
+    op.set_attr<FStatefulCompute>("FStatefulCompute<gpu>", LegacyOpForward);
     if (reg->key_var_num_args.length() != 0) {
```
Comment (eric-haibin-lin) [inline src/nnvm/legacy_op_util.cc]: How to set attr for `FStatefulComputeEx<xpu>` for legacy ops?  Do we have to modify the interface of operator class to have Forward/Backward with NDArray inputs & outputs?

Comment (piiswrong) [inline src/nnvm/legacy_op_util.cc]: Normally you use FStatefulCompute

Comment (eric-haibin-lin) [inline src/nnvm/legacy_op_util.cc]: so legacy op won't support sparse NDArrays - they'll all fallback to dense before forward/backward?

Comment (piiswrong) [inline src/nnvm/legacy_op_util.cc]: yes. You need to move to new interface to add sparse support

Comment (eric-haibin-lin) [inline src/nnvm/legacy_op_util.cc]: ok

File: include/mxnet/op_attr_types.h [hunk 2]
```diff
@@ -30,12 +92,34 @@ using nnvm::NodeAttrs;
...
+/*!
+ * \brief Execution mode of this operator.
+ */
+using FExecType = std::function<ExecType (const NodeAttrs& attrs)>;
+/*!
+ * \brief Resiger a compute function for stateful operator.
+ *
+ * \note Register under "FStatefulCompute<cpu>" and "FStatefulCompute<gpu>"
+ */
+using FStatefulCompute = std::function<void (const dmlc::any& state,
```
Comment (tqchen) [inline include/mxnet/op_attr_types.h]: const reference gives user a feeling that state won't be mutated, maybe a pointer type?

Comment (piiswrong) [inline include/mxnet/op_attr_types.h]: For mutable state you need to use shared_ptr.

Comment (tqchen) [inline include/mxnet/op_attr_types.h]: I know, but does this API implies that state can be mutated, or no?

Comment (piiswrong) [inline include/mxnet/op_attr_types.h]: state it's self cannot be mutated. If it's a pointer the content can be mutated.
The thing is state can be copied around and you cannot modify it inplace.

Another option is to use `shared_ptr<any> state` but that forces you to use a pointer and you cannot use make_shared to construct state inplace.

Comment (tqchen) [inline include/mxnet/op_attr_types.h]: This should be at least heavily documented in the comment

**Pull request merged**